lib/compiler/shared/ast.ts

```diff
@@ -0,0 +1,85 @@
+export type ParentNode = RootNode | ElementNode | ComponentNode | CustomElementNode | ExpressionNode;
+export type Node = RootNode | ElementNode | ComponentNode | CustomElementNode | ExpressionNode | TextNode | FrontmatterNode | DoctypeNode | CommentNode;
+
+export interface Position {
+  start: Point;
+  end?: Point;
+}
+export interface Point {
+  /** 1-based line number */
+  line: number;
+  /** 1-based column number, per-line */
+  column: number;
+  /** 0-based byte offset */
+  offset: number;
+}
+export interface BaseNode {
+  type: string;
+  position?: Position;
+}
+
+export interface ParentLikeNode extends BaseNode {
+  type: 'element' | 'component' | 'custom-element' | 'expression' | 'root';
+  children: Node[];
+}
+
+export interface LiteralNode extends BaseNode {
+  type: 'text' | 'doctype' | 'comment' | 'frontmatter';
+  value: string;
+}
```

Comment:
This is a pretty good approximation of our AST! Intentionally very close to [`rehype`](https://github.com/rehypejs/rehype) which is built on [`unist`](https://github.com/syntax-tree/unist).